### ISSUE TITLE
Issue with 'text/plain' and 'application/vnd.ms-excel' content_type 

### ISSUE THREAD

**Issue by jessie19** [igorkasyanchuk/active_storage_validations#291]
Hi,

Thanks for fixing content_types quickly at v1.3.1, v1.3.2, v1.3.3.

We have following 2 content types are *not accepted anymore after v1.3.0 version release, possible impact by this PR: https://github.com/igorkasyanchuk/active_storage_validations/pull/278/

Here is the model setup we have

```ruby
class FileUpload < ApplicationRecord
  enum status: {
    success: 0,
    failed: 1,
    in_progress: 2
  }

  has_one_attached :file
  # application/vnd.ms-excel needed for FireFox compatability
  # https://support.mozilla.org/en-US/questions/1401889
  validates :file, attached: true, content_type: ['text/csv', 'text/plain', 'application/vnd.ms-excel']
end
```

'text/csv' is accepted and no issues, we have an issues at another 2 content types:  'text/plain' and 'application/vnd.ms-excel'

**Comment by Mth0158** [comment]
Arf I wish it would be solved, there might be some edge cases remaining.
Can you elaborate on the attachment process (using blob, or hash, etc)? And give us a file that should but does not pass the validation?

**Comment by jessie19** [comment]
> Arf I wish it would be solved, there might be some edge cases remaining. Can you elaborate on the attachment process (using blob, or hash, etc)? And give us an example file that should but does not pass the validation?

Yeah, here is a text case

```ruby
class FileUploadTest < ActiveSupport::TestCase
  # PASS
  test "model should be invalid if file content type is image" do
    filename = 'large_image.jpeg'
    file = File.new(Rails.root.join('test', 'fixtures', 'files', filename))
    tmpfile = Tempfile.new(filename)
    tmpfile.write(file)

    upload = FileUpload.new(status: :in_progress, filename: filename)
    upload.file.attach(io: tmpfile, filename: filename)
    assert_not upload.valid?
    assert_equal ["File has an invalid content type"], upload.errors.full_messages
  end
  
  # FAILED
  test "model should be valid for FireFox MIME content type" do
    filename = 'bulk_upload_test.csv'
    file = File.new(Rails.root.join('test', 'fixtures', 'files', filename))
    tmpfile = Tempfile.new(filename)
    tmpfile.write(file)

    upload = FileUpload.new(status: :in_progress, filename: filename)
    upload.file.attach(io: tmpfile, filename: filename, content_type: 'application/vnd.ms-excel')
    assert upload.valid?
  end
end
```

**Comment by Mth0158** [comment]
Hi,
Thanks for the issue context, it helps a lot. I think I know why it fails, we have added a new check for the `content_type_validator`, it now also checks the extension match with the declared content_type.

Meaning, if you provide a file with filename `example.png` but with a declared content_type `text/plain`, it will invalidate the file since `text/plain` content type files cannot have a `.png` extension.

https://github.com/igorkasyanchuk/active_storage_validations/blob/b726c6dbe2462769d095e848b8605587d215ce11/lib/active_storage_validations/content_type_validator.rb#L69

I would suggest that you use a correct extension during your tests:
```ruby
# 1st FAIL
filename = 'bulk_upload_test.xlsx'
```
and
```ruby
# 2nd FAIL
filename = 'bulk_upload_test.txt'
```

After consideration, we should have made a major version, this is breaking too much for a "bug" fix. Sorry for the inconvenience.

**Comment by jessie19** [comment]
Hi @Mth0158

Thank you for a reply.
The reason we have content type different than the filename extension
- `filename = 'bulk_upload_test.csv'`
- content type to be `'application/vnd.ms-excel'`

It is particular for Firebox browser compatibility when selecting a CSV file, Firefox will get content type as `'application/vnd.ms-excel'`

Here is a reference
- https://support.mozilla.org/en-US/questions/1401889 

Here is a testing playground to test out on Firefox. Open console log to see content type.
- https://codepen.io/GrizzlyStu/pen/YzjZjez

But testing on FireFox 132.0.2, it seems issue is resolved, we get correct content type `'text/csv'` now.
See the below screenshot. 
<img width="1791" alt="Screenshot 2024-11-13 at 8 54 26 AM" src="https://github.com/user-attachments/assets/cd03ac9f-7268-4241-a12a-4251421405df">

We will update our product codes to meet exactly feature required and update test cases.
Close this issue, thank you s much!

**Comment by jessie19** [comment]
Sorry, after deeper investigation.

At Windows machine, CSV is associated with the **application/vnd.ms-excel** MIME type.
At Mac machine, CSV is associated with the **text/csv**.

So this maybe an edge case when a filename extension does not cover file MIMe type / content type?

**Comment by jessie19** [comment]
Here is a reference about CSV upload validation failure: a story of Windows, Chrome, and MIME Types
- https://christianwood.net/posts/csv-file-upload-validation/

**Comment by Mth0158** [comment]
@jessie19 damn that's something that I was unaware of. There is however a solution to this.

To determine the possible extensions of a file we use `Marcel::TYPE_EXTS[@attachable_content_type]`. This code is based on the `marcel` gem, which is the reference gem for content_types in Rails.

It is possible to extend the gem, we are doing it in [marcel_extensor.rb](https://github.com/igorkasyanchuk/active_storage_validations/blob/master/lib/active_storage_validations/marcel_extensor.rb). Therefore, it is possible to do something like the below to make it work for this case:
```ruby
Marcel::MimeType.extend "application/vnd.ms-excel", extensions: %(csv)
Marcel::MimeType.extend "text/plain", extensions: %(csv)
```

This will add `.csv` extension as a possible extension for `application/vnd.ms-excel` and `text/plain` content types. We could add these lines to the [marcel_extensor.rb](https://github.com/igorkasyanchuk/active_storage_validations/blob/master/lib/active_storage_validations/marcel_extensor.rb) file in the gem so everyone benefits from it.

I am creating an issue at https://github.com/rails/marcel/issues first to be sure it is a valid demand, maybe there is some content type issues lying around.

**Comment by jessie19** [comment]
Thanks @Mth0158 for very detail explanation, it helps a lot! We will follow up  https://github.com/rails/marcel/issues/116, then go from there.

Very appreciate your feedback 🙏

**Comment by Mth0158** [comment]
You are welcome! Anyway I think that we will drop the extension check in the content_type validator in the next bugfix release: https://github.com/igorkasyanchuk/active_storage_validations/blob/b726c6dbe2462769d095e848b8605587d215ce11/lib/active_storage_validations/content_type_validator.rb#L65

It is just a bad check, there are so many cases that can go wrong apparently. So you could wait till tomorrow or the day after to get the fix rather than extending `marcel` :)

**Comment by Mth0158** [comment]
Incoming bugfix release in the next hours/tomorrow to definitely fix this issue => we do not perform the extension vs content_type that causes this mess. Sorry for the unexpected breaking changes, it should have been tested more.

Let us know if you have any features idea to enhance the gem by the way :)

**Comment by Mth0158** [comment]
The fix has been released, upgrade to [1.3.4](https://github.com/igorkasyanchuk/active_storage_validations/releases/tag/1.3.4) to fix the issue :)

**Comment by jessie19** [comment]
> The fix has been released, upgrade to [1.3.4](https://github.com/igorkasyanchuk/active_storage_validations/releases/tag/1.3.4) to fix the issue :)

Huge thanks to you! It’s challenging to maintain a popular library and accommodate all use cases, so I really appreciate your quick response and swift turnaround. Thanks again @Mth0158!

**Comment by jessie19** [comment]
Confirmed no issues after using 1.3.4 version
